APE_public.rst

```diff
@@ -0,0 +1,284 @@
+   for assistive tools.
+
+Starting with PEP 8, it states that ::
+
+   Documented interfaces are considered public, unless the documentation
+   explicitly declares them to be provisional or internal interfaces exempt
+   from the usual backwards compatibility guarantees. All undocumented
+   interfaces should be assumed to be internal.
+
+   To better support introspection, modules should explicitly declare the names
+   in their public API using the ``__all__`` attribute. Setting ``__all__`` to
+   an empty list indicates that the module has no public API.
+
+   Even with ``__all__`` set appropriately, internal interfaces (packages,
+   modules, classes, functions, attributes or other names) should still be
+   prefixed with a single leading underscore.
+
+   An interface is also considered internal if any containing namespace
+   (package, module or class) is considered internal.
+
+   Imported names should always be considered an implementation detail. Other
+   modules must not rely on indirect access to such imported names unless they
+   are an explicitly documented part of the containing module’s API, such as
+   os.path or a package’s __init__ module that exposes functionality from
+   submodules.
+
+
+This is a more detailed and nuanced description of the three means discussed
+above.  It is also the most authoritative source, as it is the official Python
+style guide.  However, it is not as unambiguous as we require: which takes
+precedence, the ``__all__`` attribute or the underscore prefix?
+
+Scipy tries to resolve this ambiguity by stating that ::
+
+   - Methods / functions / classes and module attributes whose names begin with
+     a leading underscore are private.
+   - If a class name begins with a leading underscore, none of its members are
+     public, whether or not they begin with a leading underscore.
+   - If a module name in a package begins with a leading underscore none of its
+     members are public, whether or not they begin with a leading underscore.
+   - If a module or package defines ``__all__``, that authoritatively defines
+     the public interface.
+   - If a module or package doesn’t define ``__all__``, then all names that
+     don’t start with a leading underscore are public.
+
+This is a good solution, and is consistent with PEP 8, but for our goal of
+complete unambiguity and uniformity it falls short on two counts:  first, it
+does not mention the documentation; second, PEP 8 recommends "modules should
+explicitly declare the names in their public API using the ``__all__``
+attribute. Setting ``__all__`` to an empty list indicates that the module has no
+public API." SciPy allows modules to lack an ``__all__`` attribute, meaning a
+user and their tools must understand the nuances of the previous rules. Having
+an ``__all__`` attribute in every module is simpler, unambiguous, and better for
+introspection by both users and automated systems.
```

Comment:
It might be worth mentioning at this point how a user would know that a function/class is in ``__all__``. Is it not possible to explicitly import functions/classes in a module that are not in ``__all__``? If so then having something in ``__all__`` does not stop e.g. CoPilot from suggesting code that uses a function not in ``__all__``.